notebooks/e2e/e2e.py

```diff
@@ -0,0 +1,921 @@
+        )
+
+    # add a point marker to the map, if one is provided
+    if point is not None:
+        marker_kwds = {
+            "icon": Icon(
+                color="red",
+                prefix="fa",
+                icon="flag-checkered",
+            )
+        }
+        m = point.explore(
+            m=m,
+            name=point_control_name,
+            marker_type="marker",
+            marker_kwds=marker_kwds,
+            show=show_point,
+        )
+
+        # add in a dashed buffer around the point, if requested
+        if point_buffer is not None:
+            m = (
+                point.to_crs("EPSG:27700")
+                .buffer(point_buffer)
+                .explore(
+                    m=m,
+                    color=point_color,
+                    style_kwds={"fill": None, "dashArray": 5},
+                    name="Max Distance from Destination",
+                    show=show_point,
+                )
+            )
+
+    # add in an extra overlay layer, if requested
+    if overlay is not None:
+        m = overlay.explore(
+            m=m,
+            color="#F46A25",
+            name=overlay_control_name,
+        )
+
+    # get and fit the bounds to the added map layers
+    m.fit_bounds(m.get_bounds())
+
+    # add a layer control button
+    folium.LayerControl().add_to(m)
+
+    # write to file if requested
+    if save is not None:
+        dir_name = os.path.dirname(save)
+        if not os.path.exists(dir_name):
+            os.makedirs(dir_name)
+        m.save(save)
+
+    return m
+
+
+# %%
+# visualise for an ID within the UC
+UC_ID = 4110
+assert UC_ID in travel_times.to_id.unique()
+snippet_id = travel_times[travel_times.to_id == UC_ID]
+snippet_id = pop_gdf.merge(snippet_id, left_on="id", right_on="from_id")
+
+if analyse_net_config["write_outputs"]:
+    save_path = here(
+        analyse_net_config["outputs_dir"] + f"{UC_ID}_cental_uc.html"
+    )
+else:
+    save_path = None
+
+plot(
+    snippet_id,
+    uc_gdf=uc_gdf[0:1],
+    point=centroid_gdf[centroid_gdf.id == UC_ID],
+    column="travel_time",
+    column_control_name="Travel Time",
+    point_control_name="Destination Centroid",
+    caption="Median Travel Time (mins)",
+    save=save_path,
+)
+
+# %%
+# visualise for an ID within the UC
+UC_ID = 5137
+assert UC_ID in travel_times.to_id.unique()
+snippet_id = travel_times[travel_times.to_id == UC_ID]
+snippet_id = pop_gdf.merge(snippet_id, left_on="id", right_on="from_id")
+
+if analyse_net_config["write_outputs"]:
+    save_path = here(analyse_net_config["outputs_dir"] + f"{UC_ID}_sw_uc.html")
+else:
+    save_path = None
+
+plot(
+    snippet_id,
+    uc_gdf=uc_gdf[0:1],
+    point=centroid_gdf[centroid_gdf.id == UC_ID],
+    column="travel_time",
+    column_control_name="Travel Time",
+    point_control_name="Destination Centroid",
+    caption="Median Travel Time (mins)",
+    save=save_path,
+)
+
+# %%
+# visualise for an ID within the UC
+UC_ID = 3974
```

Comment:
Scenario to handle? I worked through the notebook with GHS pop2030 . I didn't get UC_ID 3974 for some reason.

`travel_times.to_id.unique()`
`…3945, 3946, 3947, 3948, 3949, 3950, 4004, 4005, 4006, 4007, 4008,…`

I have the intermediate outputs on my local machine if any of this would be useful in troubleshooting. Are the IDs consistent in time series? Is it a problem if they're not?

Reply:
This is an interesting find!

I don't think it's a scenario to handle since the IDs are purely supposed to be a unique identifier within the context of the area being explored. As long as the IDs are unique across the board, then that's all good.

The only thing I can think of that would cause changes in IDs would be the boundaries or thresholds are changed when reading in the population data. The IDs are assigned in 'order' from the top leftmost corner, so if any of the boundaries or thresholds are changed the IDs would change too. If it's not this, we should for sure explore in more detail to find the cause, since for the same boundary and thresholds the IDs should be the same.